Replace _dynamo.config with an object instead of module

qihqi opened the pull request:
Summary:
    Replace _dynamo.config with an object instead of module

    Current usage patterns of setting and reading fields on config will work
    unchanged.

    Only changes needed going forward:
    1. import torch._dynamo.config will not work. However, just doing
       import torch._dynamo is sufficient to access dynamo config
       as torch._dynamo.config.

    2. Files inside of _dynamo folder need to access config via
       from torch._dynamo.config_util import config instead of
       from torch._dynamo import config. Because _dynamo/__init__.py
       imports some of the files so it would be circular import.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #ISSUE_NUMBER


cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

yanbing-j commented:
Hi @qihqi , I suppose the `third_party/ideep` change in this PR is not intentional. Could you please help double check and remove the change? Thank you!

qihqi commented:
> 

Yes. Removed.

williamwen42 commented on torch/_dynamo/_config.py:
Is this module still needed?

qihqi replied:
no. removed.

qihqi replied:
Probly i'll do the same for inductor config; but not the same PR

williamwen42 reviewed (commented):
Will you also do the same for inductor config?

qihqi commented:
@pytorchbot merge

qihqi commented:
@pytorchbot merge

voznesenskym commented:
Why did we make this change?

jansel commented:
This will break a lot of user code (we have been telling people about dynamo.config), with seemingly little benefit.

jansel reviewed (changes requested):
This will break lot of user code with little benefit.  If there is a good reason for this change it should be done in a non-BC breaking way.

qihqi commented on test/inductor/test_config.py:
```diff
@@ -28,8 +28,6 @@ def test_set(self):
         config.max_fusion_size = 13337
         self.assertEqual(config.max_fusion_size, 13337)
         self.assertEqual(config.to_dict()["max_fusion_size"], 13337)
-        config.to_dict()["max_fusion_size"] = 32
```
Which places? I can fix up instead?

The thing is, with a dataclass the data is not backend by a flatten'd dict

qihqi replied:
Grepping didnt yield any related things:
(torch39) han@helene ~/code/pytorch $ grep 'to_dict()' torch -r
torch/_dynamo/config_utils.py:        self.to_dict().update(pickle.loads(data))
torch/_inductor/__init__.py:    current_config: Dict[str, Any] = config.to_dict()  # type: ignore[attr-defined]
torch/__init__.py:        current_config: Dict[str, Any] = config.to_dict()  # type: ignore[attr-defined]
(plus things under quantization/ which is probly a different config).

jansel replied:
Not all the users are going to be in the pytorch codebase.  One of the side effects of being an open source project.  This seems like something that is BC breaking for no reason.  And will also result in really nasty bugs for anyone relying on this since the deleted test will run fine without error, but silently do nothing.

qihqi replied:
Sure. done.

qihqi commented on torch/_dynamo/__init__.py:
```diff
@@ -1,3 +1,17 @@
+DynamoConfig.__doc__ = "Dataclass that holds configs related to dynamo."
+config = DynamoConfig()
+config.__name__ = f"{__name__}.config"
+sys.modules[f"{__name__}.config"] = config
```
sure. done

qihqi commented:
Hi @jansel any more comments? Looks like the CIs is also looking good. Thanks!

jansel commented on torch/_dynamo/allowed_functions.py:
```diff
@@ -164,7 +164,7 @@ def _is_allowed_module_prefix(obj):
             return False
 
-        mod_name = module.__name__
+        mod_name = getattr(module, "__name__", "")
```
Why is this change needed?

jansel commented on torch/_dynamo/__init__.py:
```diff
@@ -1,3 +1,14 @@
+import sys
+
+import torch._config_utils
```
Unused?

jansel commented on torch/_dynamo/__init__.py:
```diff
@@ -1,3 +1,14 @@
+import sys
```
Unused?

jansel commented on torch/_dynamo/config.py:
```diff
@@ -14,30 +15,30 @@
 
 # Assume these functions return constants
-constant_functions = {
+constant_functions: Dict[ModuleType, bool] = {
```
This type is wrong, these are functions not modules

jansel commented on torch/_dynamo/config.py:
```diff
@@ -176,63 +179,63 @@
 
 # root folder of the project
-base_dir = dirname(dirname(dirname(abspath(__file__))))
+base_dir: bool = dirname(dirname(dirname(abspath(__file__))))
```
Type is wrong here too.

jansel commented on torch/_dynamo/config.py:
```diff
@@ -133,25 +135,26 @@
 # dynamo - Runs minifier on the TorchDynamo produced graphs, if compilation fails
 # aot - Runs minifier on the Aot Autograd produced graphs, if compilation fails
-repro_after = os.environ.get("TORCHDYNAMO_REPRO_AFTER", None)
+repro_after: bool = os.environ.get("TORCHDYNAMO_REPRO_AFTER", None)
```
Type is wrong

jansel commented on torch/_functorch/__init__.py:
```diff
@@ -3,3 +3,9 @@
+from .._config_utils import make_config_dataclass
+from . import config
+
+FunctorchConfig = type(config)
```
Should this be FuncTorchConfig?

jansel commented on torch/_inductor/__init__.py:
```diff
@@ -1,8 +1,14 @@
+import sys
```
Unused?

jansel commented on torch/_inductor/__init__.py:
```diff
@@ -1,8 +1,14 @@
 from typing import Any, Dict, List, Optional
 
 import torch.fx
+from .._config_utils import make_config_dataclass
```
Unused?

jansel commented on torch/_inductor/config.py:
```diff
@@ -1,129 +1,137 @@
 
 # fallback to eager for random/dropout, this is slow but useful for debugging
-fallback_random = False
+fallback_random: int = False
```
Wrong type

jansel commented on torch/_inductor/config.py:
```diff
@@ -1,129 +1,137 @@
 
 # automatically create fallbacks when encountering an unhandled op
-implicit_fallbacks = True
+implicit_fallbacks: int = True
```
Wrong type

jansel commented on torch/_inductor/config.py:
```diff
@@ -1,129 +1,137 @@
 
 # fuse even in cases without common reads
-aggressive_fusion = False
+aggressive_fusion: int = False
```
Wrong type

jansel commented on torch/_inductor/config.py:
```diff
@@ -1,129 +1,137 @@
 
 # warnings intended for PyTorch developers, disable for point releases
-is_nightly_or_source = "dev" in torch.__version__ or "git" in torch.__version__
+is_nightly_or_source: str = "dev" in torch.__version__ or "git" in torch.__version__
```
Wrong type

jansel commented on torch/_inductor/config.py:
```diff
@@ -141,64 +149,67 @@ def decide_compile_threads():
 
 
-compile_threads = decide_compile_threads()
+def is_fbcode():
```
There is also a version of this in ConfigMixin.  Which one gets called?  Could lead to confusing bugs if one edits the wrong one.

jansel commented on torch/_inductor/config.py:
```diff
@@ -208,150 +219,153 @@ class cpp:
-
+    upload_tar: Any = dataclasses.field(default=None, metadata={"skip_pickle": True})
 
-_save_config_ignore = {
-    # workaround: "Can't pickle <function ...>"
-    "trace.upload_tar",
-}
```
Won't we still need this workaround?  I believe this was an fbcode issue.

qihqi replied:
pickle will try to pickle the `trace` object, which will then call `__getstate__` on it, `__getstate__` defined by ConfigMixin will ignore fields marked with skip_pickle.

I added a test to make sure.

jansel reviewed (changes requested):
Please double check types.  I found a lot of incorrect ones.

Please add tests, with added complexity I am worried some of the function (especially around nested configs) won't work right and arenn't being tested.

qihqi reviewed (commented):
Double checked

qihqi reviewed (commented):
Double checked

qihqi committed:
Replace _dynamo.config with an dataclass object

Summary:

Currently _dynamo.config (and friends) is written as module but
is an object with its data backed by a dictionary. This change
makes it a dataclass instead.

Nested namespace via class (`class cpp:` etc in _inductor/config.py)
are also becoming dataclass.

For now, the top level dataclass generated dynamically from the module
attr declarations. Next step is to also remove that and just use a
normal dataclass.

The goal of having dataclasses for config is that there is a name for
the config class, and therefore we can write functions that take those
configs as input.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

reset test files

qihqi reviewed (commented):
Added unit test and docstring for ConfigMixin

williamwen42 commented on torch/_dynamo/config.py:
```diff
@@ -16,30 +18,30 @@
+constant_functions: Dict[Callable, bool] = {
     torch.jit.is_scripting: False,
     torch.jit.is_tracing: False,
     torch._C._get_tracing_state: None,
```
Type mismatch between `None` and `bool`

qihqi replied:
done.

williamwen42 commented on torch/_inductor/config.py:
```diff
@@ -211,153 +214,157 @@ class cpp:
+    vec_isa_ok: Optional[str] = None
 
     # similar to config.triton.descriptive_names
     descriptive_names = "original_aten"
```
Missing annotation?

williamwen42 reviewed (commented):
LGTM other than a few nits, deferring final approval to @jansel.

jansel reviewed (changes requested):
Please remove the changes to `third_party/kineto`

qihqi commented:
> Please remove the changes to `third_party/kineto`

Done.

jansel commented:
Tests seem broken.

qihqi commented:
Thanks Jason!

Hi @williamwen42 can you delink the Meta diff with this PR so the meta internal changes check can pass? 
I'll figure out the EasyCLA thingy meanwhile. Thanks!

williamwen42 commented:
Unlinked the PR, but the meta internal check is still failing. Looped in with devinfra and was advised that this behavior is strange and that we can just merge -f once the CLA check passes.

qihqi commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
Hi @qihqi! There is a tiny bug in the changes. You [removed](https://github.com/pytorch/pytorch/pull/96455/files#diff-5f3e7e4866f93af05691e2382d2258cad959091492f900583a49e55266f5e897L5) the `tempfile` import, but it [still used in the code](https://github.com/pytorch/pytorch/pull/96455/files#diff-5f3e7e4866f93af05691e2382d2258cad959091492f900583a49e55266f5e897L249).

qihqi commented:
> Hi @qihqi! There is a tiny bug in the changes. You [removed](https://github.com/pytorch/pytorch/pull/96455/files#diff-5f3e7e4866f93af05691e2382d2258cad959091492f900583a49e55266f5e897L5) the `tempfile` import, but it [still used in the code](https://github.com/pytorch/pytorch/pull/96455/files#diff-5f3e7e4866f93af05691e2382d2258cad959091492f900583a49e55266f5e897L249).

good catch! sorry about that
